zerver/webhooks/jira/view.py

```diff
@@ -78,23 +81,59 @@ def convert_jira_markup(content: str, realm: Realm) -> str:
     full_link_re = re.compile(r"\[(?:(?P<title>[^|~]+)\|)(?P<url>[^\]]*)\]")
     content = re.sub(full_link_re, r"[\g<title>](\g<url>)", content)
 
-    # Try to convert a Jira user mention of format [~username] into a
-    # Zulip user mention. We don't know the email, just the Jira username,
-    # so we naively guess at their Zulip account using this
-    mention_re = re.compile(r"\[~(.*?)\]")
-    for username in mention_re.findall(content):
-        # Try to look up username
-        user_profile = guess_zulip_user_from_jira(username, realm)
-        if user_profile:
-            replacement = f"**{user_profile.full_name}**"
-        else:
-            replacement = f"**{username}**"
+    # We replace mentioned account [~accountid:60e5a86a471e61006a4c51fd]
+    # with their corresponding usernames fetched by Jira API.
+    url = get_url(payload)
+    if auth:
+        content = replace_account_ids_with_usernames(content, realm, url, auth)
 
-        content = content.replace(f"[~{username}]", replacement)
+    # Here's how Jira escapes special characters in their payload:
+    # ,.?\\!\n\"'\n\\[]\\{}()\n@#$%^&*\n~`|/\\\\
+    # for some reason, as of writing this, ! has two '\' before it.
+    content = content.replace("\\!", "!")
 
     return content
 
 
+def replace_account_ids_with_usernames(
+    comment: str, realm: Realm, url: str, auth: tuple[str, str]
+) -> str:
+    pattern = re.compile(r"\[~accountid:([\w\-:]+)\]")
+
+    # Use re.finditer to find all matches
+    matches = list(pattern.finditer(comment))
+
+    # Process each match in the loop
+    for match in matches:
+        account_id = match.group(1)
+        response = requests.get(f"{url}?accountId={account_id}", auth=auth)
+        data = response.json()
+        if response.status_code == 200:
+            username = data["displayName"]
+            comment = comment.replace(match.group(0), f"**{username}**")
+
+            # We also try to convert a Jira user mention of format [~accountID:DUM1234] into a
+            # Zulip user mention. We don't know the email, just the Jira username,
+            # so we naively guess at their Zulip account using this
+            user_profile = guess_zulip_user_from_jira(username, realm)
+            if user_profile:
+                replacement = f"**{user_profile.full_name}**"
+            else:
+                replacement = f"**{username}**"
+            comment = comment.replace(f"{username}", replacement)
+        else:
+            raise SuspiciousOperation("Error accessing Jira API.")
```

Comment:
I don't think we want to raise this error here. [SuspiciousOperation](https://docs.djangoproject.com/en/5.1/ref/exceptions/#suspiciousoperation) is raised when a user is performing a suspicious operation from security point of view.
>The [SuspiciousOperation](https://docs.djangoproject.com/en/5.1/ref/exceptions/#django.core.exceptions.SuspiciousOperation) exception is raised when a user has performed an operation that should be considered suspicious from a security perspective, such as tampering with a session cookie. Subclasses of SuspiciousOperation include:

We can use `JsonableError` from `zerver/lib/exceptions.py` instead.

I'm not sure why we need to do two replaces to reformat the username. 

So, all together I think we can adjust the function to this:
```python
# Process each match in the loop
    for match in matches:
        account_id = match.group(1)
        response = requests.get(f"{url}?accountId={account_id}", auth=auth)
        if response.status_code != requests.codes.ok:
            raise JsonableError(_("Error accessing Jira API."))
        data = response.json()
        username = data["displayName"]
        # We also try to convert a Jira user mention of format [~accountID:DUM1234] into a
        # Zulip user mention. We don't know the email, just the Jira username,
        # so we naively guess at their Zulip account using this
        user_profile = guess_zulip_user_from_jira(username, realm)
        if user_profile:
            username = user_profile.full_name
        comment = comment.replace(match.group(0), f"**{username}**")
    return comment
```

Reply:
Thanks for the help @PieterCK !
I'll make the changes ASAP. also I just got to know that due to `config_options` email id is undergoing double URL encoding so my email : `reharshmeena@gmail.com` is transforming into `reharshmeena%2540gmail.com` which results into failed request, also trying to figure this out.

Reply:
Ah ok, we can also do that as a follow-up work if it's too messy to deal with now.